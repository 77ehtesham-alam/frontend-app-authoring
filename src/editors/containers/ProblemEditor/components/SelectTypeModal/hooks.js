import { useEffect, useState } from 'react';
import {
  AdvanceProblemKeys, AdvanceProblems, ProblemTypeKeys, ProblemTypes,
} from '../../../../data/constants/problem';
import { StrictDict, snakeCaseKeys } from '../../../../utils';
// This 'module' self-import hack enables mocking during tests.
// See src/editors/decisions/0005-internal-editor-testability-decisions.md. The whole approach to how hooks are tested
// should be re-thought and cleaned up to avoid this pattern.
// eslint-disable-next-line import/no-self-import
import * as module from './hooks';
import { getDataFromOlx } from '../../../../data/redux/thunkActions/problem';

export const state = StrictDict({
  // eslint-disable-next-line react-hooks/rules-of-hooks
  selected: (val) => useState(val),
});

export const selectHooks = () => {
  const [selected, setSelected] = module.state.selected(ProblemTypeKeys.SINGLESELECT);
  return {
    selected,
    setSelected,
  };
};

export const onSelect = ({
  selected,
  updateField,
  setBlockTitle,
  defaultSettings,
}) => () => {
  if (Object.values(AdvanceProblemKeys).includes(selected)) {
    updateField({ problemType: ProblemTypeKeys.ADVANCED, rawOLX: AdvanceProblems[selected].template });
    setBlockTitle(AdvanceProblems[selected].title);
  } else {
    const newOLX = ProblemTypes[selected].template;
    const newState = getDataFromOlx({
      rawOLX: newOLX,
      rawSettings: {
        weight: 1,
        attempts_before_showanswer_button: 0,
        show_reset_button: null,
        showanswer: null,
        defaultToAdvanced: false,
      },
      defaultSettings: snakeCaseKeys(defaultSettings),
    });
    updateField(newState);
    setBlockTitle(ProblemTypes[selected].title);
  }
};

export const useArrowNav = (selected, setSelected) => {
  const detectKeyDown = (e) => {
    const problemTypeValues = Object.values(ProblemTypeKeys);
    switch (e.key) {
      case 'ArrowUp':
        if (problemTypeValues.includes(selected) && ProblemTypes[selected].prev) {
          setSelected(ProblemTypes[selected].prev);
          document.getElementById(ProblemTypes[selected].prev).focus();
        }
        break;
      case 'ArrowDown':
        if (problemTypeValues.includes(selected) && ProblemTypes[selected].next) {
          setSelected(ProblemTypes[selected].next);
          document.getElementById(ProblemTypes[selected].next).focus();
        }
        break;
      default:
    }
  };
  useEffect(() => {
    document.addEventListener('keydown', detectKeyDown, true);
    return () => {
      document.removeEventListener('keydown', detectKeyDown, true);
    };
  }, [selected, setSelected]);
};
