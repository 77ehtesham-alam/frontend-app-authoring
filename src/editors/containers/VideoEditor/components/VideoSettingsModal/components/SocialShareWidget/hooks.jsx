import { useSelector } from 'react-redux';
import analyticsEvents from './constants';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';
import { selectors } from '../../../../../../data/redux';

export const handleSocialSharingCheckboxChange = ({ updateField }) => {
  const analytics = useSelector(selectors.app.analytics);
  const allowVideoSharing = useSelector(selectors.video.allowVideoSharing);
  return (event) => {
    sendTrackEvent(
      analyticsEvents.socialSharingSettingChanged,
      {
        ...analytics,
        value: event.target.checked,
      }
    );
    updateField({
      allowVideoSharing: {
        ...allowVideoSharing,
        value: event.target.checked,
      },
    });
  };
};
