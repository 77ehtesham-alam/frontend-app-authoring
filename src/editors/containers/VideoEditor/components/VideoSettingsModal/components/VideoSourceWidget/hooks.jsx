import { actions } from '../../../../../../data/redux';
import { parseYoutubeId } from '../../../../../../data/services/cms/api';
import * as requests from '../../../../../../data/redux/thunkActions/requests';

export const sourceHooks = ({ dispatch }) => ({
  updateVideoURL: (e, videoId) => {
    const videoUrl = e.target.value;
    dispatch(actions.video.updateField({ videoSource: videoUrl }));

    const youTubeId = parseYoutubeId(videoUrl);
    if (youTubeId) {
      dispatch(requests.checkTranscriptsForImport({
        videoId,
        youTubeId,
        onSuccess: (response) => {
          if (response.data.command === 'import') {
            dispatch(actions.video.updateField({
              allowTranscriptImport: true,
            }));
          }
        },
      }));
    }
  },
  updateVideoId: (e) => dispatch(actions.video.updateField({ videoId: e.target.value })),
});

export const fallbackHooks = ({ fallbackVideos, dispatch }) => ({
  addFallbackVideo: () => dispatch(actions.video.updateField({ fallbackVideos: [...fallbackVideos, ''] })),
  deleteFallbackVideo: (videoUrl) => {
    const updatedFallbackVideos = fallbackVideos.splice(fallbackVideos.indexOf(videoUrl), 1);
    dispatch(actions.video.updateField({ fallbackVideos: updatedFallbackVideos }));
  },
});

export default {
  sourceHooks,
  fallbackHooks,
};
