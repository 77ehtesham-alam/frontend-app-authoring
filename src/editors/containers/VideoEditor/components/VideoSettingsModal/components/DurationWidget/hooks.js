import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';

import { actions, selectors } from '../../../../../../data/redux';
import messages from '../messages';

import * as module from './hooks';

const durationMatcher = /^(\d{0,2}):?(\d{0,2})?:?(\d{0,2})?$/i;

export const durationWidget = ({ dispatch }) => {
  const reduxStartStopTimes = useSelector(selectors.video.duration);
  const setReduxStartStopTimes = (val) => dispatch(actions.video.updateField({ duration: val }));
  const initialState = module.durationString(reduxStartStopTimes);
  const [unsavedStartStopTimes, setUnsavedStartStopTimes] = useState(initialState);

  useEffect(() => {
    setUnsavedStartStopTimes(module.durationString(reduxStartStopTimes));
  }, [reduxStartStopTimes]);

  return {
    reduxStartStopTimes,
    unsavedStartStopTimes,
    onBlur: (index) => (
      (e) => module.updateDuration({
        reduxStartStopTimes,
        setReduxStartStopTimes,
        unsavedStartStopTimes,
        setUnsavedStartStopTimes,
        index,
        durationString: e.target.value,
      })
    ),
    onChange: (index) => (
      (e) => setUnsavedStartStopTimes(module.onDurationChange(unsavedStartStopTimes, index, e.target.value))
    ),
    onKeyDown: (index) => (
      (e) => setUnsavedStartStopTimes(module.onDurationKeyDown(unsavedStartStopTimes, index, e))
    ),
    getTotalLabel: ({ duration, subtitle, intl }) => {
      if (!duration.stopTime) {
        if (!duration.startTime) {
          return intl.formatMessage(messages.fullVideoLength);
        }
        if (subtitle) {
          return intl.formatMessage(
            messages.startsAt,
            { startTime: module.durationStringFromValue(duration.startTime) },
          );
        }
        return null;
      }
      const total = duration.stopTime - (duration.startTime || 0);
      return intl.formatMessage(messages.total, { total: module.durationStringFromValue(total) });
    },
  };
};

/**
 * durationString(duration)
 * Returns the display value for embedded start and stop times
 * @param {object} duration - object containing startTime and stopTime millisecond values
 * @return {object} - start and stop time from incoming object mapped to duration strings.
 */
export const durationString = (duration) => ({
  startTime: module.durationStringFromValue(duration.startTime),
  stopTime: module.durationStringFromValue(duration.stopTime),
});

/**
 * durationStringFromValue(value)
 * Returns a duration string in 'hh:mm:ss' format from the given ms value
 * @param {number} value - duration (in milliseconds)
 * @return {string} - duration in 'hh:mm:ss' format
 */
export const durationStringFromValue = (value) => {
  // return 'why';
  if (!value || typeof value !== 'number' || value <= 0) {
    return '00:00:00';
  }
  const seconds = Math.floor((value / 1000) % 60);
  const minutes = Math.floor((value / 60000) % 60);
  const hours = Math.floor((value / 3600000) % 60);
  const zeroPad = (num) => String(num).padStart(2, '0');
  return [hours, minutes, seconds].map(zeroPad).join(':');
};

/**
 * updateDuration({ reduxStartStopTimes, unsavedStartStopTimes, setUnsavedStartStopTimes, setReduxStartStopTimes })
 * Returns a memoized callback based on inputs that updates unsavedStartStopTimes value and form value
 * if the new string is valid (reduxStartStopTimes stores a number, unsavedStartStopTimes stores a string).
 * If the duration string is invalid, resets the unsavedStartStopTimes value to the latest good value.
 * @param {object} reduxStartStopTimes - redux-stored durations in milliseconds
 * @param {object} unsavedStartStopTimes - hook-stored duration in 'hh:mm:ss' format
 * @param {func} setReduxStartStopTimes - set form value
 * @param {func} setUnsavedStartStopTimes - set unsavedStartStopTimes object
 * @param {string} index - startTime or stopTime
 * @return {func} - callback to update duration unsavedStartStopTimesly and in redux
 *   updateDuration(args)(index, durationString)
 */
export const updateDuration = ({
  reduxStartStopTimes,
  unsavedStartStopTimes,
  setReduxStartStopTimes,
  setUnsavedStartStopTimes,
  index,
  inputString,
}) => {
  let newDurationString = inputString;
  let newValue = module.valueFromDuration(newDurationString);
  // maxTime is 23:59:59 or 86399 seconds
  if (newValue > 86399000) {
    newValue = 86399000;
  }
  // stopTime must be at least 1 second, if not zero
  if (index === 'stopTime' && newValue > 0 && newValue < 1000) {
    newValue = 1000;
  }
  // stopTime must be at least 1 second after startTime, except 0 means no custom stopTime
  if (index === 'stopTime' && newValue > 0 && newValue < (reduxStartStopTimes.startTime + 1000)) {
    newValue = reduxStartStopTimes.startTime + 1000;
  }
  // startTime must be at least 1 second before stopTime, except when stopTime is less than a second
  // (stopTime should only be less than a second if it's zero, but we're being paranoid)
  if (index === 'startTime' && reduxStartStopTimes.stopTime >= 1000 && newValue > (reduxStartStopTimes.stopTime - 1000)) {
    newValue = reduxStartStopTimes.stopTime - 1000;
  }
  newDurationString = module.durationStringFromValue(newValue);
  setUnsavedStartStopTimes({ ...unsavedStartStopTimes, [index]: newDurationString });
  setReduxStartStopTimes({ ...reduxStartStopTimes, [index]: newValue });
};

/**
 * onDurationChange(duration)
 * Returns a new duration value based on onChange event
 * @param {object} duration - object containing startTime and stopTime millisecond values
 * @param {string} index - 'startTime or 'stopTime'
 * @param {string} val - duration in 'hh:mm:ss' format
 * @return {object} duration - object containing startTime and stopTime millisecond values
 */
export const onDurationChange = (duration, index, val) => {
  const match = val.trim().match(durationMatcher);
  if (!match) {
    return duration;
  }

  const caretPos = document.activeElement.selectionStart;
  let newDuration = val;
  if (caretPos === newDuration.length && (newDuration.length === 2 || newDuration.length === 5)) {
    newDuration += ':';
  }

  return {
    ...duration,
    [index]: newDuration,
  };
};

/**
 * onDurationKeyDown(duration)
 * Returns a new duration value based on onKeyDown event
 * @param {object} duration - object containing startTime and stopTime millisecond values
 * @param {string} index - 'startTime or 'stopTime'
 * @param {Event} event - event from onKeyDown
 * @return {object} duration - object containing startTime and stopTime millisecond values
 */
export const onDurationKeyDown = (duration, index, event) => {
  const caretPos = document.activeElement.selectionStart;
  let newDuration = duration[index];

  switch (event.key) {
    case 'Enter':
      document.activeElement.blur();
      break;
    case 'Backspace':
      if (caretPos === newDuration.length && newDuration.slice(-1) === ':') {
        newDuration = newDuration.slice(0, -1);
      }
      break;
    default:
      break;
  }

  return {
    ...duration,
    [index]: newDuration,
  };
};

/**
 * valueFromDuration(duration)
 * Returns a millisecond duration value from the given 'hh:mm:ss' format string
 * @param {string} duration - duration in 'hh:mm:ss' format
 * @return {number} - duration in milliseconds. Returns null if duration is invalid.
 */
export const valueFromDuration = (duration) => {
  let matches = duration.trim().match(durationMatcher);
  if (!matches) {
    return 0;
  }
  matches = matches.slice(1).filter(v => v !== undefined);
  if (matches.length < 3) {
    for (let i = 0; i <= 3 - matches.length; i++) {
      matches.unshift(0);
    }
  }
  const [hours, minutes, seconds] = matches.map(x => parseInt(x, 10) || 0);
  return ((hours * 60 + minutes) * 60 + seconds) * 1000;
};

export default {
  durationWidget,
  durationString,
  durationStringFromValue,
  updateDuration,
  onDurationChange,
  onDurationKeyDown,
  valueFromDuration,
};
