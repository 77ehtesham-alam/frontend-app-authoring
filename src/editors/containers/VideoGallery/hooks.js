import React from 'react';
import * as module from './hooks';
import messages from './messages';
import {
  filterKeys,
  filterMessages,
  sortKeys,
  sortMessages,
} from './utils';

export const state = {
  highlighted: (val) => React.useState(val),
  searchString: (val) => React.useState(val),
  showSelectVideoError: (val) => React.useState(val),
  showSizeError: (val) => React.useState(val),
  sortBy: (val) => React.useState(val),
  filertBy: (val) => React.useState(val),
  hideSelectedVideos: (val) => React.useState(val),
};

export const searchAndSortHooks = () => {
  const [searchString, setSearchString] = module.state.searchString('');
  const [sortBy, setSortBy] = module.state.sortBy(sortKeys.dateNewest);
  const [filterBy, setFilterBy] = module.state.filertBy(filterKeys.videoStatus);
  const [hideSelectedVideos, setHideSelectedVideos] = module.state.hideSelectedVideos(false);

  return {
    searchString,
    onSearchChange: (e) => setSearchString(e.target.value),
    clearSearchString: () => setSearchString(''),
    sortBy,
    onSortClick: (key) => () => setSortBy(key),
    sortKeys,
    sortMessages,
    filterBy,
    onFilterClick: (key) => () => setFilterBy(key),
    filterKeys,
    filterMessages,
    showSwitch: true,
    hideSelectedVideos,
    switchMessage: messages.hideSelectedCourseVideosSwitchLabel,
    onSwitchClick: () => setHideSelectedVideos(!hideSelectedVideos),
  };
};

export const videoListHooks = ({ videos }) => {
  const [highlighted, setHighlighted] = module.state.highlighted(null);
  const [
    showSelectVideoError,
    setShowSelectVideoError,
  ] = module.state.showSelectVideoError(false);
  const [
    showSizeError,
    setShowSizeError,
  ] = module.state.showSizeError(false);
  const filteredList = videos; // TODO missing filters and sort
  return {
    galleryError: {
      show: showSelectVideoError,
      set: () => setShowSelectVideoError(true),
      dismiss: () => setShowSelectVideoError(false),
      message: messages.selectVideoError,
    },
    // TODO We need to update this message when implementing the video upload screen
    inputError: {
      show: showSizeError,
      set: () => setShowSizeError(true),
      dismiss: () => setShowSelectVideoError(false),
      message: messages.fileSizeError,
    },
    galleryProps: {
      galleryIsEmpty: Object.keys(filteredList).length === 0,
      searchIsEmpty: filteredList.length === 0,
      displayList: filteredList,
      highlighted,
      onHighlightChange: (e) => setHighlighted(e.target.value),
      emptyGalleryLabel: messages.emptyGalleryLabel,
      showIdsOnCards: true,
      height: '100%',
    },
    selectBtnProps: {
      onclick: () => {
        // TODO Update this when implementing the selection feature
      },
    },
  };
};

export const fileInputHooks = () => {
  // TODO [Update video] Implement this
  const ref = React.useRef();
  const click = () => ref.current.click();

  return {
    click,
    addFile: () => {},
    ref,
  };
};

export const filterAssets = ({ assets }) => {
  let videos = [];
  const assetsList = Object.values(assets);
  if (assetsList.length > 0) {
    videos = assetsList.filter(asset => asset?.contentType?.startsWith('video/'));
  }
  return videos;
};

export const videoHooks = ({ videos }) => {
  const searchSortProps = module.searchAndSortHooks();
  const videoList = module.videoListHooks({ videos });
  const {
    galleryError,
    galleryProps,
    inputError,
    selectBtnProps,
  } = videoList;
  const fileInput = module.fileInputHooks();

  return {
    galleryError,
    inputError,
    fileInput,
    galleryProps,
    searchSortProps,
    selectBtnProps,
  };
};

export default {
  videoHooks,
  filterAssets,
};
