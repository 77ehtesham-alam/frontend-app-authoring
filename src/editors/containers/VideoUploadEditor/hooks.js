import * as requests from '../../data/redux/thunkActions/requests';
import * as module from './hooks';
import { selectors } from '../../data/redux';
import store from '../../data/store';
import * as appHooks from '../../hooks';

export const {
  navigateTo,
} = appHooks;

export const uploadVideo = async ({ dispatch, supportedFiles }) => {
  const data = { files: [] };
  supportedFiles.forEach((file) => {
    data.files.push({
      file_name: file.name,
      content_type: file.type,
    });
  });
  const onFileUploadedHook = module.onFileUploaded();
  dispatch(await requests.uploadVideo({
    data,
    onSuccess: async (response) => {
      const { files } = response.data;
      await Promise.all(Object.values(files).map(async (fileObj) => {
        const fileName = fileObj.file_name;
        const uploadUrl = fileObj.upload_url;
        const edxVideoId = fileObj.edx_video_id;
        const uploadFile = supportedFiles.find((file) => file.name === fileName);

        // TODO I added this temporally to test the redirecton without
        // make the post to the upload URL. I added this also after the success post
        // To test this I overwriten my own response with an existing edx_video_id on
        // the edx-platform view: https://github.com/openedx/edx-platform/blob/master/cms/djangoapps/contentstore/views/videos.py#L224
        onFileUploadedHook(edxVideoId);

        if (!uploadFile) {
          console.error(`Could not find file object with name "${fileName}" in supportedFiles array.`);
          return;
        }
        const formData = new FormData();
        formData.append('uploaded-file', uploadFile);
        await fetch(uploadUrl, {
          method: 'PUT',
          body: formData,
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        })
          .then(() => onFileUploadedHook(edxVideoId))
          .catch((error) => console.error('Error uploading file:', error));
      }));
    },
  }));
};

export const onFileUploaded = () => {
  const state = store.getState();
  const learningContextId = selectors.app.learningContextId(state);
  const blockId = selectors.app.blockId(state);
  return (edxVideoId) => navigateTo(`/course/${learningContextId}/editor/video/${blockId}?selectedVideoId=${edxVideoId}`);
};

export default {
  uploadVideo,
};
