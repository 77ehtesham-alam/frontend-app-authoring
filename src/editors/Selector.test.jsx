import React from 'react';
import { useDispatch } from 'react-redux';
import { shallow } from 'enzyme';
import * as hooks from './hooks';
import Selector from './Selector';

jest.mock('./hooks', () => ({
  initializeApp: jest.fn(),
}));

jest.mock('./containers/VideoGallery', () => 'VideoGallery');

const props = {
  learningContextId: 'course-v1:edX+DemoX+Demo_Course',
  lmsEndpointUrl: 'evenfakerurl.com',
  studioEndpointUrl: 'fakeurl.com',
};

const initData = {
  blockId: '',
  blockType: 'video',
  ...props,
};

describe('Editor', () => {
  describe('render', () => {
    test('rendering correctly with expected Input', () => {
      expect(shallow(<Selector {...props} />)).toMatchSnapshot();
    });
  });
  describe('behavior', () => {
    it('calls initializeApp hook with dispatch, and passed data', () => {
      shallow(<Selector {...props} />);
      expect(hooks.initializeApp).toHaveBeenCalledWith({
        dispatch: useDispatch(),
        data: initData,
      });
    });
  });
});
