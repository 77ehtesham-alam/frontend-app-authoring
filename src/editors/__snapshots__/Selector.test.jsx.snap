// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor render rendering correctly with expected Input 1`] = `<VideoGallery />`;
